{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def logo():\n",
    "    print(f'\\033[1;32;40m{\"-\"* 100}')\n",
    "    print(f'\\033[1;32;40m{\"-\"* 100}\\n')\n",
    "\n",
    "\n",
    "\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 3) and column > 0 and column < 5) or ((column == 5 or column == 1) and (row == 1 or row == 2))):  \n",
    "                result_str=result_str+\"p\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 6) and (column > 1 and column < 5)) or (column == 5 and row != 0 and row != 6)):  \n",
    "                result_str=result_str+\"D\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 6 or row ==3) and (column > 1 and column < 5)) or (column == 5 and row != 0 and row != 6)):  \n",
    "                result_str=result_str+\"B\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    print(\"-\"* 100)\n",
    "    print(\"-\"* 100)\n",
    "\n",
    "logo()\n",
    "\n",
    "three2one = {\n",
    "    'ALA' : 'A',    \n",
    "    'ARG' : 'R',    \n",
    "    'ASN' : 'N',    \n",
    "    'ASP' : 'D',    \n",
    "    'CYS' : 'C',    \n",
    "    'GLU' : 'E',    \n",
    "    'GLN' : 'Q',    \n",
    "    'GLY' : 'G',    \n",
    "    'HIS' : 'H',    \n",
    "    'ILE' : 'I',    \n",
    "    'LEU' : 'L',    \n",
    "    'LYS' : 'K',    \n",
    "    'MET' : 'M',    \n",
    "    'PHE' : 'F',    \n",
    "    'PRO' : 'P',    \n",
    "    'SER' : 'S',    \n",
    "    'THR' : 'T',    \n",
    "    'TRP' : 'W',    \n",
    "    'TYR' : 'Y',   \n",
    "    'VAL' : 'V'\n",
    "}\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code used to prototype the program "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary to convert three letter protein code to single letter\n",
    "# it is used in the write and alignment functions\n",
    "three2one = {\n",
    "    'ALA' : 'A',    \n",
    "    'ARG' : 'R',    \n",
    "    'ASN' : 'N',    \n",
    "    'ASP' : 'D',    \n",
    "    'CYS' : 'C',    \n",
    "    'GLU' : 'E',    \n",
    "    'GLN' : 'Q',    \n",
    "    'GLY' : 'G',    \n",
    "    'HIS' : 'H',    \n",
    "    'ILE' : 'I',    \n",
    "    'LEU' : 'L',    \n",
    "    'LYS' : 'K',    \n",
    "    'MET' : 'M',    \n",
    "    'PHE' : 'F',    \n",
    "    'PRO' : 'P',    \n",
    "    'SER' : 'S',    \n",
    "    'THR' : 'T',    \n",
    "    'TRP' : 'W',    \n",
    "    'TYR' : 'Y',   \n",
    "    'VAL' : 'V'\n",
    "}\n",
    "\n",
    "\n",
    "user_input = input(\"Enter the path to your file: \")\n",
    "\n",
    "\n",
    "import sys\n",
    "import os\n",
    "import Bio\n",
    "\n",
    "#the read option that allows a user to read-in PDB files\n",
    "def read():   \n",
    "    import sys\n",
    "    import os\n",
    "    import urllib.request\n",
    "    \n",
    "    assert os.path.exists(user_input)\n",
    "\n",
    "    src = \"http://www.rcsb.org/pdb/downloadFile.do\" \\\n",
    "        + \"?fileFormat=pdb&compression=NO&structureId=%s\"\n",
    "\n",
    "    #The pdb files you want to read are recorded in a text file(PDBfiles.txt)\n",
    "    with open(\"PDBfiles.txt\",\"w\") as fh:\n",
    "        filename= input(\"Enter PDB filename to be readin:\\n\")\n",
    "        filename=filename.upper()\n",
    "        fh.write(filename)\n",
    "\n",
    "    #The pdb files are readin and stored in the PDBfiles directory    \n",
    "    with open(\"PDBfiles.txt\",\"r\") as fh:\n",
    "        for line in fh:\n",
    "            pdb = line\n",
    "\n",
    "        filename = os.path.join(user_input, pdb+\".pdb\")\n",
    "\n",
    "        if os.path.isfile(filename) :\n",
    "                print (\"Already read in \" + pdb)\n",
    "                main\n",
    "        else :\n",
    "            print (\"Downloading \" + pdb)\n",
    "            print(\"Please wait\")\n",
    "            urllib.request.urlretrieve( src % pdb, filename)\n",
    "            print (\"Successfully read in\")\n",
    "    \n",
    "            \n",
    "\n",
    "#the search function allows the user to search for gylcosylation sites\n",
    "# an example of a PDB file with glycosylation sites according to the pattern is 4AQD.\n",
    "# you can read_in the PDB file and search for the glycosylation sites\n",
    "def search():\n",
    "    import re\n",
    "    Entries=os.listdir(user_input)\n",
    "    print('Files already read in:')\n",
    "    for entry in Entries:\n",
    "        print(entry)\n",
    "\n",
    "    filename =  input('Choose one file to work on:\\n')\n",
    "    raw_sequence=[]\n",
    "    seff=[]  \n",
    "\n",
    "    with open(user_input+filename,'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"SEQRES\"):\n",
    "                list=(str.split(line[19:70]))\n",
    "                for res in list:\n",
    "                    seff.append(res)\n",
    "\n",
    "                    raw_sequence.append(three2one[res])\n",
    "    sequence=str(''.join(raw_sequence).replace(' ',''))\n",
    "    pattern = re.compile(r'[GYL]A[PFW][TLVGMT]') #Find patterns starting with G or Y or L then A then P or F or W\n",
    "    matches = pattern.finditer(sequence)\n",
    "    for match in matches:\n",
    "        print(match)\n",
    "\n",
    "#the write function allows you to write the various section in files\n",
    "#you may write to a .txt file or a .fasta file\n",
    "def write():\n",
    "    from Bio.Seq import Seq\n",
    "    from Bio.SeqRecord import SeqRecord\n",
    "    from Bio.Alphabet import IUPAC\n",
    "    Entries=os.listdir(user_input)\n",
    "    print('Files already read in:')\n",
    "    for entry in Entries:\n",
    "        print(entry)\n",
    "    filename =  input('Choose one file to work on:\\n')\n",
    "    \n",
    "    options = input(\"\"\"\n",
    "    1-Write out coordinate file\n",
    "    2-Write out sequence (Fasta format)\n",
    "    \n",
    "    \"\"\")\n",
    "    \n",
    "    if options== \"1\":\n",
    "        sub_options=input(\"\"\"\n",
    "        a- All atom\n",
    "        b- Backbone atoms\n",
    "        c-Alpha carbon atoms only\n",
    "        \"\"\")\n",
    "        \n",
    "        if sub_options == \"a\":\n",
    "            input_file= input(\"Please enter a .txt file to write the coordinates:\\n\")\n",
    "            with open(user_input+filename,'r') as f:\n",
    "                with open(input_file,\"w\") as fh:\n",
    "                    for line in f:\n",
    "                        if line.startswith(\"ATOM\"):\n",
    "                            fh.write(line)\n",
    "                    print(f'Coordinates written succesfully in: {input_file}')\n",
    "\n",
    "        if sub_options == \"b\":\n",
    "            input_file= input(\"Please enter a .txt file to write the coordinates:\\n\")\n",
    "            with open(user_input+filename,'r') as f:\n",
    "                with open(input_file,\"w\") as fh:\n",
    "                    for line in f:\n",
    "                        if line[:4] == 'ATOM' and line[12:16] == \" CA \":\n",
    "                            fh.write(line)\n",
    "                        elif line[:4] == 'ATOM' and line[12:15] == \" N \":\n",
    "                            fh.write(line)\n",
    "                        elif line[:4] == 'ATOM' and line[12:15] == \" C \":\n",
    "                            fh.write(line)\n",
    "                        elif line[:4] == 'ATOM' and line[12:15] == \" O \":\n",
    "                            fh.write(line)\n",
    "                    print(f'Coordinates written succesfully in: {input_file}')\n",
    "\n",
    "        if sub_options == \"c\":\n",
    "            input_file= input(\"Please enter a .txt file to write the coordinates:\\n\")\n",
    "            with open(user_input+filename,'r') as f:\n",
    "                with open(input_file,\"w\") as fh:\n",
    "                    for line in f:\n",
    "                        if line[:4] == 'ATOM' and line[12:16] == \" CA \":\n",
    "                            fh.write(line)\n",
    "                    print(f'Coordinates written succesfully in: {input_file}')\n",
    "                        \n",
    "    if options== \"2\":\n",
    "         sub_options=input(\"\"\"\n",
    "    1-SEQRES Sequence\n",
    "    2-Coordinate Sequence\n",
    "    3-Alignment Sequence\n",
    "    \"\"\")\n",
    "    \n",
    "    \n",
    "    if sub_options == \"1\":\n",
    "        input_file= input(\"Please enter a .fasta file to write the SEQRES sequence:\\n\")\n",
    "        input_file= input(\"Please enter a .fasta file to write the SEQRES sequence:\\n\")\n",
    "        with open(user_input+filename,'r') as f:\n",
    "            with open(input_file,\"w\") as fh:\n",
    "                print(\"Converting to FASTA...\")\n",
    "                print(\"Just smile...\")\n",
    "                print(\"Please wait...\")\n",
    "                for strLine in f:\n",
    "                    if strLine.startswith(\"SEQRES\"):\n",
    "                        line=strLine[17:]\n",
    "                        strLine=line.rstrip(\"\\n\")\n",
    "                        record = SeqRecord(Seq(strLine,IUPAC.protein),id=\"PK_025292.1\", name=filename, \n",
    "                                           description=f'PDB files in {user_input}')\n",
    "                        fh.write(record.format('fasta'))\n",
    "\n",
    "                print(\"Written succesfully\")\n",
    "                    \n",
    "    if sub_options == \"2\":\n",
    "        input_file= input(\"Please enter a .fasta file to write the Coordinate sequence:\\n\")    \n",
    "        with open(user_input+filename,'r') as f:\n",
    "            with open(input_file,\"w\") as fh:\n",
    "                print(\"Converting to FASTA...\")\n",
    "                print(\"Just smile...\")\n",
    "                print(\"Please wait...\")\n",
    "                for strLine in f:\n",
    "                    if strLine.startswith(\"ATOM\"):\n",
    "                        line=strLine[16:20]\n",
    "                        strLine=line.rstrip(\"\\n\")\n",
    "                        record = SeqRecord(Seq(strLine,IUPAC.protein),id=\"PK_025292.1\", name=filename, \n",
    "                                           description=f'PDB files in {user_input}')\n",
    "                        fh.write(record.format('fasta'))\n",
    "\n",
    "                print(\"Written succesfully\")\n",
    "                \n",
    "    if sub_options == \"3\":\n",
    "        input_file= input(\"Please enter a .txt file to write the Alignment sequence:\\n\") \n",
    "        raw_sequence=[]\n",
    "        seff=[]  \n",
    "\n",
    "        with open(user_input+filename,'r') as f:\n",
    "            with open(input_file,\"w\") as fh:\n",
    "                for line in f:\n",
    "                    if line.startswith(\"SEQRES\"):\n",
    "                        list=(str.split(line[19:70]))\n",
    "                        for res in list:\n",
    "                            seff.append(res)\n",
    "\n",
    "                            raw_sequence.append(three2one[res])\n",
    "                sequence=str(''.join(raw_sequence).replace(' ',''))\n",
    "                record = SeqRecord(Seq(sequence,IUPAC.protein),id=\"PK_025292.1\", name=filename, \n",
    "                                           description=f'PDB files in {user_input}')\n",
    "                fh.write(record.format('fasta'))\n",
    "\n",
    "                print(\"Written succesfully\")\n",
    "                \n",
    "                \n",
    "\n",
    "#this function gives the display of the write function options without writing the output to a file\n",
    "def information() :   \n",
    "    from Bio.Seq import Seq\n",
    "    from Bio.SeqRecord import SeqRecord\n",
    "    from Bio.Alphabet import IUPAC\n",
    "    Entries=os.listdir(user_input)\n",
    "    print('Files already read in:')\n",
    "    for entry in Entries:\n",
    "        print(entry)\n",
    "    filename =  input('Choose one file to work on:\\n')\n",
    "\n",
    "    sub_options=input(\"\"\"\n",
    "    1-Display coordinate sequence (as explained above)\n",
    "    2-Display SEQRES sequence (as explained above)\n",
    "    3-Display Alignment sequence (as explained above)\n",
    "    4-Display all non-water ligands in the protein (if any)\n",
    "\n",
    "    \"\"\")\n",
    "\n",
    "    if sub_options==\"1\":\n",
    "        with open(user_input+filename,'r') as f:\n",
    "            for line in f:\n",
    "                if line.startswith(\"ATOM\"):\n",
    "                    print(line[30:-27])\n",
    "\n",
    "    if sub_options==\"2\":\n",
    "        with open(user_input+filename,'r') as f:\n",
    "            for line in f:\n",
    "                if line.startswith(\"SEQRES\"):\n",
    "                    line=line[17:].rstrip()\n",
    "                    print(line)\n",
    "\n",
    "    if sub_options==\"3\":\n",
    "        raw_sequence=[]\n",
    "        seff=[]  \n",
    "        with open(user_input+filename,'r') as f:\n",
    "            for line in f:\n",
    "                if line.startswith(\"SEQRES\"):\n",
    "                    list=(str.split(line[19:70]))\n",
    "                    for res in list:\n",
    "                        seff.append(res)\n",
    "\n",
    "                        raw_sequence.append(three2one[res])\n",
    "            sequence=str(''.join(raw_sequence).replace(' ',''))\n",
    "            record = SeqRecord(Seq(sequence,IUPAC.protein),id=\"PK_025292.1\", name=filename, \n",
    "                                       description=f'PDB files in {user_input}')\n",
    "            print(record)\n",
    "            \n",
    "            \n",
    "\n",
    "#this function allows the user to choose two files of his/her choice which have been read in and do an alignment\n",
    "#the alignment is written into a file which the user chooses\n",
    "#this function uses biopython to do a pairwise global alignment\n",
    "#since the sequences may be too long the program only aligns the first 50 amino acids but can be changed give a wider coverage\n",
    "def alignment():\n",
    "    raw_sequence1=[]\n",
    "    seff1=[] \n",
    "    raw_sequence2=[]\n",
    "    seff2=[]\n",
    "    \n",
    "    Entries=os.listdir(user_input)\n",
    "    print('Files already read in:')\n",
    "    for entry in Entries:\n",
    "        print(entry)\n",
    "    print(\"Please select two PDB files to be aligned\")\n",
    "    \n",
    "    filename1=input(\"Enter the name of the first file:\\n\")\n",
    "    filename2=input(\"Enter the name of the second file:\\n\")\n",
    "    output_file=input(\"Enter a .txt file to write the alignment to:\\n\")\n",
    "    \n",
    "    with open(user_input+filename1,'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"SEQRES\"):\n",
    "                list=(str.split(line[19:70]))\n",
    "                for res in list:\n",
    "                    seff1.append(res)\n",
    "\n",
    "                    raw_sequence1.append(three2one[res])\n",
    "    sequence1=str(''.join(raw_sequence1).replace(' ',''))\n",
    "    #print(sequence1)\n",
    "    X=sequence1[:50]\n",
    "                   \n",
    "                \n",
    "    with open(user_input+filename2,'r') as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"SEQRES\"):\n",
    "                list=(str.split(line[19:70]))\n",
    "                for res in list:\n",
    "                    seff2.append(res)\n",
    "\n",
    "                    raw_sequence2.append(three2one[res])\n",
    "    sequence2=str(''.join(raw_sequence2).replace(' ',''))\n",
    "    #print(sequence2)\n",
    "    Y=sequence2[:50]\n",
    "                               \n",
    "    # Import pairwise2 module\n",
    "    from Bio import pairwise2\n",
    "\n",
    "    # Import format_alignment method\n",
    "    from Bio.pairwise2 import format_alignment\n",
    "\n",
    "    # Define two sequences to be aligned\n",
    "\n",
    "    #X \n",
    "    #Y\n",
    "\n",
    "    # Get a list of the global alignments between the two sequences ACGGGT and ACG satisfying the given scoring\n",
    "    # A match score is the score of identical chars, else mismatch score.\n",
    "    # Same open and extend gap penalties for both sequences.\n",
    "    alignments = pairwise2.align.globalms(X, Y, 2, -1, -0.5, -0.1)\n",
    "\n",
    "    # Use format_alignment method to format the alignments in the list\n",
    "    with open(output_file,\"w\") as fh:\n",
    "        for a in alignments:\n",
    "            fh.write(format_alignment(*a))\n",
    "    print(f'Alignment Written succesfully to {output_file}')\n",
    "\n",
    "#this function acts as a navigation tools to the user as he/she uses the program\n",
    "def help():\n",
    "    print(\"These are the various options in the programme:\")\n",
    "    print(\"\"\"\n",
    "    R – Read in PDB files\n",
    "    S – Search option\n",
    "    W – Write out option\n",
    "    I – Information option\n",
    "    A – Alignment option\n",
    "    H – Help option\n",
    "    Q – quit\n",
    "\n",
    "    \"\"\")\n",
    "    \n",
    "    \n",
    "    print(\"\"\"\n",
    "    1.The read option allows you do download any PDB file which is stored in the system to be used later on\n",
    "    \n",
    "    2.The search option allow you to find all the gycosylation sites in the PDB file\n",
    "    \n",
    "    3.The write option allow you to parse through the file and write the desired sequences into files of\n",
    "    your choice\n",
    "    \n",
    "    4.The information option displays the desired sequences without writing them to files\n",
    "    \n",
    "    5.The alignment option allows you to extract sequences from two PDB files, align them and write the \n",
    "    results into a file of your choice\n",
    "    \n",
    "    6.The help options displays for you all the options for easier navigation through the program\n",
    "    \n",
    "    7.The quit option terminates the program\n",
    "    \n",
    "    \n",
    "    \"\"\")\n",
    "    \n",
    "#this function allows you to exit the program once you are done     \n",
    "def quit():\n",
    "    print(\"Goodbye....\")\n",
    "    os._exit\n",
    "\n",
    "\n",
    "        \n",
    "        \n",
    "#this function displays the PDB program logo \n",
    "def logo():\n",
    "    print(f'\\033[1;32;40m{\"-\"* 100}')\n",
    "    print(f'\\033[1;32;40m{\"-\"* 100}\\n')\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 3) and column > 0 and column < 5) or ((column == 5 or column == 1) and (row == 1 or row == 2))):  \n",
    "                result_str=result_str+\"p\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 6) and (column > 1 and column < 5)) or (column == 5 and row != 0 and row != 6)):  \n",
    "                result_str=result_str+\"D\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    result_str=\"\";    \n",
    "    for row in range(0,7):    \n",
    "        for column in range(0,7):     \n",
    "            if (column == 1 or ((row == 0 or row == 6 or row ==3) and (column > 1 and column < 5)) or (column == 5 and row != 0 and row != 6)):  \n",
    "                result_str=result_str+\"B\"    \n",
    "            else:      \n",
    "                result_str=result_str+\" \"    \n",
    "        result_str=result_str+\"\\n\"    \n",
    "    print(f'\\033[1;32;40m{result_str}');\n",
    "\n",
    "    print(\"-\"* 100)\n",
    "    print(\"-\"* 100)   \n",
    "\n",
    "#this function merges all the options into a single function\n",
    "    def menu():\n",
    "        option=input(\"\"\"\n",
    "        R – Read in PDB files\n",
    "        S – Search option\n",
    "        W – Write out option\n",
    "        I – Information option\n",
    "        A – Alignment option\n",
    "        H – Help option\n",
    "        Q – quit\n",
    "\n",
    "        \"\"\")\n",
    "        if option !='Q':\n",
    "            if option.upper() in ('R','S','W','I','A','H'):\n",
    "                if option.upper() == 'R':\n",
    "                    read()\n",
    "                    menu()\n",
    "                if option.upper() == 'S':\n",
    "                    search()\n",
    "                    menu()\n",
    "                if option.upper() == 'W':\n",
    "                    write()\n",
    "                    menu()\n",
    "                if option.upper() == 'I':\n",
    "                    information()\n",
    "                    menu()\n",
    "                if option.upper() == 'A':\n",
    "                    alignment()\n",
    "                    menu()\n",
    "                if option.upper() == 'H':\n",
    "                    help()\n",
    "                    menu()\n",
    "                if option.upper() == 'Q':\n",
    "                    quit()\n",
    "    menu()\n",
    "logo()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "user_input = input(\"Enter the path of your file: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import tempfile\n",
    "filename= input(\"file:\\n\")\n",
    "pdb= tempfile.TemporaryFile()\n",
    "\n",
    "with open('pdb','w') as fh:\n",
    "    fh.write(filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
